OpenSim/Common/Mtx.cpp

```diff
@@ -34,214 +35,27 @@
-        rP[0] =  c*p0 + s*p2;
-        rP[1] =  p1;
-        rP[2] = -s*p0 + c*p2;
-
-    // Z
-    } else if(aXYZ==2) {
-        rP[0] =  c*p0 - s*p1;
-        rP[1] =  s*p0 + c*p1;
-        rP[2] =  p2;
-    }
-}
-//_____________________________________________________________________________
-/**
- * Rotate a point about a specified axis by a specified amount.
- *
- * @param aAxis Axis about which to rotate.  The axis is not assumed to be
- * a unit vector; however, if the axis is the zero vector, no rotation
- * is performed.
- * @param aRadians Amount of rotation in radians.
- * @param aP Point to rotate.
- * @param rP Rotated point.  aP and rP may coincide in memory.
- */
-void Mtx::
-Rotate(const double aAxis[3],double aRadians,const double aP[3],double rP[3])
-{
-    if(aAxis==0) return;
-    if(aP==NULL) return;
-    if(rP==NULL) return;
-
-}
-
-//-----------------------------------------------------------------------------
-// ROTATE BY DEGREES
-//-----------------------------------------------------------------------------
-//_____________________________________________________________________________
-/**
- * Rotate a point about the X axis by a specified amount.
- *
- * @param aXYZ Specify whether to rotate about the X (aXYZ=0), Y (aXYZ=1),
- * or Z (aXYZ=2) axis.  If aXYZ is not 0, 1, or 2, no rotation is performed.
- * @param aDegrees Amount of rotation in degrees.
- * @param aP Point to rotate.
- * @param rP Rotated point.  aP and rP may coincide in memory.
- */
-void Mtx::
-RotateDeg(int aXYZ,double aDegrees,const double aP[3],double rP[3])
-{
-    Rotate(aXYZ,SimTK_DEGREE_TO_RADIAN*aDegrees,aP,rP);
-}
-//_____________________________________________________________________________
-/**
- * Rotate a point about a specified axis by a specified amount.
- *
- * @param aAxis Axis about which to rotate.
- * @param aDegrees Amount of rotation in degrees.
- * @param aP Point to rotate.
- * @param rP Rotated point.  aP and rP may coincide in memory.
- */
-void Mtx::
-RotateDeg(const double aAxis[3],double aDegrees,const double aP[3],double rP[3])
-{
-    Rotate(aAxis,SimTK_DEGREE_TO_RADIAN*aDegrees,aP,rP);
-}
-
+/*static*/ thread_local int Mtx::_PSpaceSize = 0;
+/*static*/ thread_local int Mtx::_WSpaceSize = 0;
+/*static*/ thread_local double** Mtx::_P1Space = NULL;
+/*static*/ thread_local double** Mtx::_P2Space = NULL;
+/*static*/ thread_local double*  Mtx::_WSpace = NULL;
 
+static const double eps = std::numeric_limits<double>::epsilon();
 //=============================================================================
 // MATRIX ARITHMATIC
 //=============================================================================
 //_____________________________________________________________________________
 /**
- * Assign a square matrix to the identidy matrix:  rI = I.
+ * Assign a square matrix to the identity matrix:  rI = I.
  * The matrix must be square.
  *
+ * //TODO: Is there a better way to do this? Seems kind of inefficient
```

Comment:
For contiguous memory, I think the best way to create an identity matrix is to zero the n^2 memory as fast as possible (probably with `memset()`), then go back and fill in the n diagonals (as a 1d vector with stride n). That avoids having a conditional in the innermost loop.

Even better, use `SimTK::Matrix mat(n,n); mat=1.;` which will create an nxn identity matrix using that approach.